aws/resource_aws_emr_cluster_test.go

```diff
@@ -3397,3 +3440,176 @@ EOF
+        size                 = 100
+        type                 = "gp2"
+        volumes_per_instance = 1
+      }
+      instance_type     = "m4.xlarge"
+      weighted_capacity = 1
+    }
+    instance_type_configs {
+      bid_price_as_percentage_of_on_demand_price = 100
+      ebs_config {
+        size                 = 100
+        type                 = "gp2"
+        volumes_per_instance = 1
+      }
+      instance_type     = "m4.2xlarge"
+      weighted_capacity = 2
+    }
+    launch_specifications {
+      spot_specification {
+		allocation_strategy      = "capacity-optimized"
+        block_duration_minutes   = 0
+        timeout_action           = "SWITCH_TO_ON_DEMAND"
+        timeout_duration_minutes = 10
+      }
+    }
+    name                      = "core fleet"
+    target_on_demand_capacity = 0
+    target_spot_capacity      = 2
+  }
+  service_role = aws_iam_role.emr_service.arn
+  depends_on   = [
+    aws_route_table_association.test,
+    aws_iam_role_policy_attachment.emr_service,
+    aws_iam_role_policy_attachment.emr_instance_profile,
+  ]
+
+  ec2_attributes {
+    subnet_id                         = aws_subnet.test.id
+    emr_managed_master_security_group = aws_security_group.test.id
+    emr_managed_slave_security_group  = aws_security_group.test.id
+    instance_profile                  = aws_iam_instance_profile.emr_instance_profile.arn
+  }
+
+  bootstrap_action {
+    path = "s3://elasticmapreduce/bootstrap-actions/run-if"
+    name = "runif"
+    args = ["instance.isMaster=true", "echo running on master node"]
+  }
+}
+`, r),
+	)
+}
+
+func testAccAWSEmrClusterConfigInstanceFleetsMasterOnly(r string) string {
+	return testAccAWSEmrComposeConfig(false,
+		testAccAWSEmrClusterConfigIAMServiceRoleBase(r),
+		testAccAWSEmrClusterConfigIAMInstanceProfileBase(r),
+		testAccAWSEmrClusterConfigBootstrapActionBucket(r),
+		fmt.Sprintf(`
+resource "aws_iam_role_policy_attachment" "emr_service_spot" {
+  role       = aws_iam_role.emr_service.id
+  policy_arn = aws_iam_policy.emr_service_spot.arn
+}
+
+resource "aws_iam_policy" "emr_service_spot" {
+  name = "%[1]s_emr_spot"
+
+  policy = <<EOT
+{
+    "Version": "2012-10-17",
+    "Statement": [{
+        "Effect": "Allow",
+        "Resource": "arn:aws:iam::*:role/aws-service-role/spot.amazonaws.com/AWSServiceRoleForEC2Spot*",
+        "Action": [
+            "iam:CreateServiceLinkedRole"
+        ]
+    }]
+}
+EOT
+}
+
+resource "aws_emr_cluster" "tf-test-cluster" {
+  name                 = "%[1]s"
+  release_label        = "emr-5.30.1"
+  applications         = ["Hadoop", "Hive"]
+  log_uri              = "s3n://terraform/testlog/"
+
+  master_instance_fleet    {
+    instance_type_configs        {
+          instance_type = "m3.xlarge"
+        }
+    
+      target_on_demand_capacity = 1
+    }
+  service_role = aws_iam_role.emr_service.arn
+  depends_on   = [
+    aws_route_table_association.test,
+    aws_iam_role_policy_attachment.emr_service,
```

Comment:
To help prevent the following error:

```
=== CONT  TestAccAWSEMRCluster_instance_fleet
TestAccAWSEMRCluster_instance_fleet: resource_aws_emr_cluster_test.go:1335: Step 1/1 error: terraform failed: exit status 1
stderr:
Error: Error waiting for EMR Cluster state to be "WAITING" or "RUNNING": TERMINATING: VALIDATION_ERROR: Service role arn:aws:iam::*******:role/tf-acc-test-8032042948779895989_default_role has insufficient EC2 permissions
--- FAIL: TestAccAWSEMRCluster_instance_fleet (96.64s)
```

We either need to include the new policy attachment as an explicit dependency here:

```
    aws_iam_role_policy_attachment.emr_service,
    aws_iam_role_policy_attachment.emr_service_spot,
```

Or better yet, we should switch the test configurations to use the automatically updated AWS managed IAM policy with all the service permissions (including this new create service linked role permission) by:

* In `testAccAWSEmrClusterConfigIAMServiceRoleBase()`:
  * Remove `aws_iam_policy` resource
  * Switch `aws_iam_role_policy_attachment` `policy_arn` to `arn:aws:iam::aws:policy/service-role/AmazonElasticMapReduceRole`
* In these new instance fleet test configurations:
  * Remove `aws_iam_policy` and `aws_iam_role_policy_attachment` resources

There are definitely EC2 permissions missing since the hardcoded policy was last updated in this file so it will be good to get that fixed.

Reply:
can I just use the default EMR role   `EMR_DefaultRole` and `EMR_EC2_DefaultRole` for the test that not related to IAM policy?

Reply:
fixed, sorry my company aws account blocked vpc and other resource permission so I cannot run the full acc test there. I have to remove all the dependency in the test case when I run acc test with my aws account.